File: backend/app/api/model_admin.py
import logging

from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException, Query
from google.cloud.firestore_v1.async_client import AsyncClient

from app.core.auth_middleware import require_auth
from app.core.firestore import COLLECTIONS, get_firestore_client
from app.services.learning_service import train_scoring_model

logger = logging.getLogger("suas.api.model_admin")

router = APIRouter(prefix="/model-admin", dependencies=[Depends(require_auth)])

TRAINING_THRESHOLD = 200


@router.get("/status")
async def get_model_status(
    db: AsyncClient = Depends(get_firestore_client),
) -> dict:
    """Return the currently active ML model's status, or a flag indicating none exists."""
    try:
        query = (
            db.collection(COLLECTIONS["model_training"])
            .where("is_active", "==", True)
            .limit(1)
        )
        async for doc in query.stream():
            data = doc.to_dict()
            return {
                "has_active_model": True,
                "model_version": data.get("model_version"),
                "r_squared": data.get("r_squared"),
                "training_set_size": data.get("training_set_size"),
                "trained_at": data.get("trained_at"),
                "gcs_path": data.get("gcs_path"),
                "top_features": data.get("top_features", []),
            }
        return {"has_active_model": False}
    except Exception as exc:
        logger.error("Failed to fetch model status: %s", exc)
        raise HTTPException(status_code=500, detail="Failed to fetch model status")


@router.get("/overrides")
async def get_overrides(
    limit: int = Query(default=20, ge=1, le=100),
    db: AsyncClient = Depends(get_firestore_client),
) -> list[dict]:
    """Return the last N override log entries, sorted by logged_at descending."""
    try:
        ref = db.collection(COLLECTIONS["model_training"]).document("overrides")
        doc = await ref.get()
        if not doc.exists:
            return []

        records: list[dict] = doc.to_dict().get("records", [])
        records.sort(key=lambda r: r.get("logged_at", ""), reverse=True)
        return records[:limit]
    except Exception as exc:
        logger.error("Failed to fetch overrides: %s", exc)
        raise HTTPException(status_code=500, detail="Failed to fetch overrides")


@router.post("/train", status_code=202)
async def trigger_training(
    background_tasks: BackgroundTasks,
    db: AsyncClient = Depends(get_firestore_client),
) -> dict:
    """Manually trigger model training in the background."""
    background_tasks.add_task(train_scoring_model, db)
    return {"status": "accepted", "message": "Training queued"}


@router.get("/post-count")
async def get_post_count(
    db: AsyncClient = Depends(get_firestore_client),
) -> dict:
    """Return how many posts are eligible for training."""
    try:
        eligible = 0
        query = db.collection(COLLECTIONS["posts"]).where(
            "status", "in", ["published", "metrics_synced"]
        )
        async for doc in query.stream():
            data = doc.to_dict()
            scoring = data.get("scoring") or {}
            if scoring.get("alignment_score") is not None:
                eligible += 1

        return {
            "eligible_posts": eligible,
            "threshold": TRAINING_THRESHOLD,
            "ready_to_train": eligible >= TRAINING_THRESHOLD,
        }
    except Exception as exc:
        logger.error("Failed to count eligible posts: %s", exc)
        raise HTTPException(status_code=500, detail="Failed to count eligible posts")

File: frontend/src/api/model_admin.ts
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query'
import { apiClient } from './client'

export interface ModelStatus {
  has_active_model: boolean
  model_version?: string
  r_squared?: number
  training_set_size?: number
  trained_at?: string
  gcs_path?: string
  top_features?: Array<{ name: string; importance: number }>
}

export interface OverrideRecord {
  post_id: string
  fernando_rating: number
  predicted_engagement: number
  normalized_rating: number
  normalized_pred: number
  divergence: number
  logged_at: string
}

export interface PostCount {
  eligible_posts: number
  threshold: number
  ready_to_train: boolean
}

export const modelAdminKeys = {
  status: ['model-admin', 'status'] as const,
  overrides: (limit: number) => ['model-admin', 'overrides', limit] as const,
  postCount: ['model-admin', 'post-count'] as const,
}

export function useModelStatus() {
  return useQuery({
    queryKey: modelAdminKeys.status,
    queryFn: async () => {
      const { data } = await apiClient.get<ModelStatus>('/api/model-admin/status')
      return data
    },
    staleTime: 60_000,
  })
}

export function useModelOverrides(limit = 20) {
  return useQuery({
    queryKey: modelAdminKeys.overrides(limit),
    queryFn: async () => {
      const { data } = await apiClient.get<OverrideRecord[]>('/api/model-admin/overrides', { params: { limit } })
      return data
    },
    staleTime: 2 * 60_000,
  })
}

export function usePostCount() {
  return useQuery({
    queryKey: modelAdminKeys.postCount,
    queryFn: async () => {
      const { data } = await apiClient.get<PostCount>('/api/model-admin/post-count')
      return data
    },
    staleTime: 60_000,
  })
}

export function useTriggerTraining() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: () => apiClient.post('/api/model-admin/train'),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: modelAdminKeys.status })
      qc.invalidateQueries({ queryKey: modelAdminKeys.postCount })
    },
  })
}
